src/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -0,0 +1,558 @@
+                var tcs = new TaskCompletionSource<int>(this);
+                BeginSend(buffers, socketFlags, iar =>
+                {
+                    var innerTcs = (TaskCompletionSource<int>)iar.AsyncState;
+                    try { innerTcs.TrySetResult(((Socket)innerTcs.Task.AsyncState).EndSend(iar)); }
+                    catch (Exception e) { innerTcs.TrySetException(e); }
+                }, tcs);
+                return tcs.Task;
+            }
+
+            // Configure the buffer list.  We don't clear the buffers when returning the SAEA to the pool,
+            // so as to minimize overhead if the same buffers are used for subsequent operations (which is likely).
+            // But SAEA doesn't support having both a buffer and a buffer list configured, so clear out a buffer
+            // if there is one before we set the desired buffer list.
+            if (saea.Buffer != null) saea.SetBuffer(null, 0, 0);
+            saea.BufferList = buffers;
+            saea.SocketFlags = socketFlags;
+
+            // Initiate the send
+            Task<int> t;
+            if (!SendAsync(saea))
+            {
+                // The operation completed synchronously.  Get a task for it and return the SAEA for future use.
+                t = saea.SocketError == SocketError.Success ?
+                    GetSuccessTask(saea) :
+                    Task.FromException<int>(new SocketException((int)saea.SocketError));
+                ReturnSocketAsyncEventArgs(saea, isReceive: false);
+            }
+            else
+            {
+                // The operation completed asynchronously.  Get the task for the operation,
+                // with appropriate synchronization to coordinate with the async callback
+                // that'll be completing the task.
+                t = saea.GetTaskSafe();
+            }
+            return t;
+        }
+
+        internal Task<int> SendToAsync(ArraySegment<byte> buffer, SocketFlags socketFlags, EndPoint remoteEP)
+        {
+            var tcs = new TaskCompletionSource<int>(this);
+            BeginSendTo(buffer.Array, buffer.Offset, buffer.Count, socketFlags, remoteEP, iar =>
+            {
+                var innerTcs = (TaskCompletionSource<int>)iar.AsyncState;
+                try { innerTcs.TrySetResult(((Socket)innerTcs.Task.AsyncState).EndSendTo(iar)); }
+                catch (Exception e) { innerTcs.TrySetException(e); }
+            }, tcs);
+            return tcs.Task;
+        }
+
+        /// <summary>Validates the supplied array segment, throwing if its array or indices are null or out-of-bounds, respectively.</summary>
+        private static void ValidateBuffer(ArraySegment<byte> buffer)
+        {
+            if (buffer.Array == null)
+            {
+                throw new ArgumentNullException(nameof(buffer.Array));
+            }
+            if (buffer.Offset < 0 || buffer.Offset > buffer.Array.Length)
```

Comment:
I'm curious, are the checks for Offset and Count really useful?  I'm not sure how you could get into this scenario without serious hackage (e.g. unsafe code).  And if you're doing that, you could very easily hack the Array pointer to point into space.  Do we gain anything by this?

Also, I notice we don't do the same validation in the BufferList case.